### PR TITLE
Improve custom pickler/unpickler generators

### PR DESCRIPTION
**Pull request by jvican**
This PR improves how custom pickler/unpickler generators are created in both
scala pickling and third-party libraries that want to define their own for
concrete types.

It also improves error reporting significantly. Now we have concrete exceptions
for every kind of error that may happen. This improvement helps readability and
also gives a pretty straightforward idea of what could go wrong.

List of changes:

Add custom runtime generators for Either

```
* Add custom runtime generators for Either and tests checking the new
 custom runtime generation.

* Make `AnyPicklerUnpickler` deal with more corner cases.

* Modify PicklingProtocol in sandbox-test and reuse it for both formats.

* Add error for disabled runtime generation.

* Improve the generator helper, make some renamings and add more
 documentation.
```

Refactor runtime custom picklers

```
* Remove some todos and make a better distinction between the runtime
 custom picklers and the runtime pickler registry. Merge the latter one
 in the `DefaultRuntimePicklerGenerator`.

* Improve some docs w.r.t. the runtime custom picklers.

* Make the `RuntimePicklerRegistry` more general for future custom
 picklers (like Either).

* Remove annoying prints when running tests.
```

Improve error reporting

```
* Keep compatibility with the same old API but improve significantly the
 way we report errors factoring out the most important ones and
 changing some of the error messages.

* Factor out the functionality to register pickler generators (to be
 used at runtime) and make easier to add new ones for the missing types.

* In tests, changes make sure that scala pickling is imported and that the
 deprecated method `generate` for Pickler and Unpickler is not used.

* Apply minor changes throughout scala pickling.
```

### CODE REVIEW

File: core/src/main/scala/scala/pickling/PicklingErrors.scala [hunk 1]
```diff
@@ -0,0 +1,165 @@
+package scala.pickling
+
+private[pickling] object Feedback {
+
+  def failedGenerationMsg(x: String, y: String, z: String) =
```
Comment (jsuereth) [inline core/src/main/scala/scala/pickling/PicklingErrors.scala:5]: can we use different arg names than x/y/z?   I'll be confused everytime I look at this method :)

Comment (jsuereth) [inline core/src/main/scala/scala/pickling/PicklingErrors.scala:5]: x= tpe, z = lookedFor, y = goal

File: core/src/main/scala/scala/pickling/PicklingErrors.scala [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    )
+
+  /** Represent any error related to parsing. */
+  class ParsingException(msg: String) extends BasePicklingException(msg)
+
+  /** Exception thrown when the parsing of a message is not successful.
+    *
+    * @param value Message that gave the error
+    */
+  final case class JsonParseFailed(value: String)
```
Comment (jsuereth) [inline core/src/main/scala/scala/pickling/PicklingErrors.scala:75]: This should probably be in the `json` library, similar for binary. 

Theoretically the json + binary formats are "standalone libs" (or could be).

File: core/src/main/scala/scala/pickling/binary/BinaryPickle.scala [hunk 3]
```diff
@@ -188,16 +188,19 @@ class BinaryPickleReader(in: BinaryInput, format: BinaryPickleFormat) extends Ab
           case NULL_TAG =>
             FastTypeTag.Null
           case ELIDED_TAG =>
-            hints.elidedType.getOrElse(throw new PicklingException(s"Type is elided in pickle, but no elide hint was provided by unpickler!"))
+            hints.elidedType.getOrElse(throw NoTypeHint)
```
Comment (jsuereth) [inline core/src/main/scala/scala/pickling/binary/BinaryPickle.scala:191]: I'm already liking the new exceptions :)

File: core/src/main/scala/scala/pickling/generator/sourcegen.scala [hunk 4]
```diff
@@ -316,7 +319,7 @@ private[pickling]  trait SourceGenerator extends Macro with tags.FastTypeTagMacr
     val tpe = x.parent.tpe[c.universe.type](c.universe)
     val defaultCase =
       if(x.lookupRuntime) CaseDef(Ident(nme.WILDCARD), EmptyTree, q"_root_.scala.pickling.internal.`package`.currentRuntime.picklers.genUnpickler(_root_.scala.pickling.internal.`package`.currentMirror, tagKey)")
-      else CaseDef(Ident(nme.WILDCARD), EmptyTree, q"""throw new _root_.scala.pickling.PicklingException("Cannot unpickle, Unexpected tag: " + tagKey + " not recognized.")""")
+      else CaseDef(Ident(nme.WILDCARD), EmptyTree, q"""throw $basePicklingException("Cannot unpickle, Unexpected tag: " + tagKey + " not recognized.")""")
```
Comment (jsuereth) [inline core/src/main/scala/scala/pickling/generator/sourcegen.scala:322]: This probably deserves to be its own exception type.  Can do later though.

File: core/src/main/scala/scala/pickling/internal/AutoDefaultRuntimePicklers.scala [hunk 5]
```diff
@@ -1,68 +0,0 @@
...
-/** Helper method to help register picklers which can handle special/crazy classes for runtime generration.
-  *
-  *
-  * Note: Currently this only handles Tuple2s
-  */
-trait RuntimePicklerRegistryHelper extends PicklerRegistry {
-  final def autoRegisterDefaults(): Unit = {
-    val tpes = List(
-      "scala.Tuple2",
-      "scala.Tuple2$mcII$sp",
```
Comment (jsuereth) [inline core/src/main/scala/scala/pickling/internal/AutoDefaultRuntimePicklers.scala:17]: I assume we still handle this hairy beasts somewhere?  If so, awesome.

File: core/src/main/scala/scala/pickling/internal/DefaultRuntimePicklerGenerator.scala [hunk 6]
```diff
@@ -21,7 +20,9 @@ import scala.reflect.runtime.{universe => ru}
 [error] 	scala.pickling.array.binary.ArrayBinaryTest
 [error] 	scala.pickling.test.runtime.array.RuntimeArrayTest
  */
-class DefaultRuntimePicklerGenerator(reflectionLock: ReentrantLock) extends spi.RuntimePicklerGenerator {
+class DefaultRuntimePicklerGenerator(reflectionLock: ReentrantLock)
```
Comment (jsuereth) [inline core/src/main/scala/scala/pickling/internal/DefaultRuntimePicklerGenerator.scala:23]: Just an FYI, I'll be altering this interface before 0.11.x goes out.  Still working out the kinks locally.

File: core/src/main/scala/scala/pickling/internal/NoSharingRefRegistry.scala [hunk 7]
```diff
@@ -29,7 +29,7 @@ final class NoSharingRefUnpicklingRegistry extends RefUnpicklingRegistry {
   override def clear(): Unit = ()
 
   /** Registers an object to an id, after its FIRST deserialization. */
-  override def regsiterUnpicklee(oid: Int, value: Any): Unit = ()
+  override def registerUnpicklee(oid: Int, value: Any): Unit = ()
```
Comment (jsuereth) [inline core/src/main/scala/scala/pickling/internal/NoSharingRefRegistry.scala:32]: these typos are embarassing, sorry!

File: core/src/main/scala/scala/pickling/internal/RuntimePicklerRegistry.scala [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+import scala.pickling.runtime.CustomRuntime
+import scala.pickling.spi.PicklerRegistry
+
+/** Helper class to register picklers and unpicklers that
+  * handle special/crazy classes for runtime generation.
+  *
+  * Related to [[scala.pickling.runtime.CustomRuntime]].
+  * Note: Currently this only handles Tuple2s.
+  */
+trait RuntimePicklerRegistry extends CustomRuntime {
```
Comment (jsuereth) [inline core/src/main/scala/scala/pickling/internal/RuntimePicklerRegistry.scala:13]: Just an FYI -

I think giving `FastTypeTag.makeRaw(clazz)` "Special powers" to handle specialized instances might actually remove the need for this entire hackery.

So, if `makeRaw` can take a specialized class but dive into the unspecialized version for generating a runtime type tag, we could actually handle a LOT more cases than just Tuple's and most/all of this code disappears.

File: core/src/main/scala/scala/pickling/pickler/Any.scala [hunk 9]
```diff
@@ -12,32 +12,44 @@ import scala.reflect.runtime.currentMirror
...
-    val clazz = picklee.getClass
-    val classLoader = this.getClass.getClassLoader
-    GRL.lock()
-    val tag = try FastTypeTag.makeRaw(clazz)
-    finally GRL.unlock()
-    val p = currentRuntime.picklers.genPickler(classLoader, clazz, tag)
-    p.asInstanceOf[Pickler[Any]].pickle(picklee, builder)
+
+    // Use nullPickler if null, get pickler otherwise
+    val pickler = if (picklee == null) nullPickler else {
```
Comment (jsuereth) [inline core/src/main/scala/scala/pickling/pickler/Any.scala:30]: ouch, good catch. I think all the generated pickling code automatically does this, so this wasn't caught be "normal" code. I think i ran into it before though, so good to see this.

File: core/src/main/scala/scala/pickling/pickler/Any.scala [hunk 10]
```diff
@@ -12,32 +12,44 @@ import scala.reflect.runtime.currentMirror
...
+      val clazz = picklee.getClass
+      val classLoader = this.getClass.getClassLoader
+      GRL.lock()
```
Comment (jsuereth) [inline core/src/main/scala/scala/pickling/pickler/Any.scala:33]: Random aside,

With the FasttypeTag refactoring, we no longer have to lock the GRL when we construct FastTypeTags.  I must have missed a bit of code where that can be fixed.  I'll open a ticket.

File: core/src/main/scala/scala/pickling/pickler/Either.scala [hunk 11]
```diff
@@ -1,69 +1,180 @@
...
+  /** Pickle as [[Any]] because the scala pickling
+    * use [[Any]] as a placeholder for the type params.
+    */
+  private def pickleAsAny[S <: Either[_, _]]
+      (picklee: S, fullTpe: FastTypeTag[S],
+       name: String, field: Any, builder: PBuilder) = {
+
+    builder.beginEntry(picklee, fullTpe)
+    builder.putField(name, { b =>
+      AnyPicklerUnpickler.pickle(field, b)
```
Comment (jsuereth) [inline core/src/main/scala/scala/pickling/pickler/Either.scala:106]: Just a note, it would be feasible to actually NOT force runtime either to use Any given the way type tags work.  i.e. you could generate the Either pickler with concrete tags.

This is probably fine as-is though :)

File: core/src/main/scala/scala/pickling/pickler/GeneratorRegistry.scala [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+  def generateHelper[T](elementTagExtractor: FastTypeTagSpecializer[T])
+                       (tpe: FastTypeTag[_]): (Pickler[T], Unpickler[T]) = {
+
+    val elementType = elementTagExtractor(tpe)
+    getPickler(elementType, tpe) -> getUnpickler(elementType, tpe)
+
+  }
+
+  /** Specialize a [[FastTypeTag]] for a type [[T]]. */
+  def specialize[T](tag: FastTypeTag[_]): FastTypeTag[T] =
```
Comment (jsuereth) [inline core/src/main/scala/scala/pickling/pickler/GeneratorRegistry.scala:73]: I see you're learning the hackery of existentials :)

File: core/src/main/scala/scala/pickling/pickler/Iterable.scala [hunk 13]
```diff
@@ -1,124 +1,135 @@
...
         TravPickler.oneArgumentTagExtractor(tpe)
-      } _
-    currentRuntime.picklers.registerPicklerUnpicklerGenerator("scala.collection.immutable.List", generator)
-    currentRuntime.picklers.registerPicklerUnpicklerGenerator("scala.collection.immutable.$colon$colon", generator)
-    // Note: this is kind of wrong/backward.  We should just directly deserialize Nil
-    currentRuntime.picklers.registerPicklerUnpicklerGenerator("scala.collection.immutable.Nil.type", generator)
+      }
+
+    // TODO: Deserialize directly Nil and remove this hack.
+    registerGen("scala.collection.immutable.Nil.type", generator)
```
Comment (jsuereth) [inline core/src/main/scala/scala/pickling/pickler/Iterable.scala:32]: This is a nice cleanup.

File: sandbox-test/src/test/scala/scala/pickling/PicklingProtocol.scala [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+import scala.pickling.binary.BinaryFormats
+import scala.pickling.internal.HybridRuntime
+import scala.pickling.json.JsonFormats
+import scala.pickling.pickler.AllPicklers
+
+trait PicklingProtocol extends {
+  val oldRuntime = internal.currentRuntime
+  val onlyLookup = {
+    val currentRuntime =
+      if(oldRuntime.isInstanceOf[HybridRuntime]) oldRuntime
```
Comment (jsuereth) [inline sandbox-test/src/test/scala/scala/pickling/PicklingProtocol.scala:12]: Is this working out ok with two protocols?  Or is it working by accident?

**Comment by jsuereth** [comment]
This looks pretty great, nice work!  A few comments (mostly useless), but a few questions as well.  Nothing that has to be handled now.

File: core/src/main/scala/scala/pickling/internal/AutoDefaultRuntimePicklers.scala (hunk 5 above)
Comment (jvican) [inline core/src/main/scala/scala/pickling/internal/AutoDefaultRuntimePicklers.scala:17]: Yes, we do :)

File: core/src/main/scala/scala/pickling/internal/RuntimePicklerRegistry.scala (hunk 8 above)
Comment (jvican) [inline core/src/main/scala/scala/pickling/internal/RuntimePicklerRegistry.scala:13]: That's very cool! Hope to see something soon :smile: Yet I think we'll need the custom runtime picklers/unpicklers to allow external third-party users to define and use them.

File: core/src/main/scala/scala/pickling/pickler/Any.scala (hunk 9 above)
Comment (jvican) [inline core/src/main/scala/scala/pickling/pickler/Any.scala:30]: Yes, probably it would be better to remove the null checks in higher layers of the interface and just reuse this :smiley:

File: core/src/main/scala/scala/pickling/pickler/Any.scala (hunk 10 above)
Comment (jvican) [inline core/src/main/scala/scala/pickling/pickler/Any.scala:33]: Cool, actually we should remove these locks because they're not going to work in scalajs (when it's possible).

File: core/src/main/scala/scala/pickling/pickler/Either.scala (hunk 11 above)
Comment (jvican) [inline core/src/main/scala/scala/pickling/pickler/Either.scala:106]: Actually, yes, I was assuming that the type parameters were Any, good catch! I'll address this in another PR.

File: sandbox-test/src/test/scala/scala/pickling/PicklingProtocol.scala (hunk 14 above)
Comment (jvican) [inline sandbox-test/src/test/scala/scala/pickling/PicklingProtocol.scala:12]: Actually, I haven't tested it yet, since we're using only `JsonPicklingProtocol`. Any special reason why this wouldn't work?

File: core/src/main/scala/scala/pickling/PicklingErrors.scala (hunk 1 above)
Comment (jvican) [inline core/src/main/scala/scala/pickling/PicklingErrors.scala:5]: I'll address this in a follow-up PR. Sorry! I was very haskell-ish (using a lazy-strategy).

File: core/src/main/scala/scala/pickling/PicklingErrors.scala (hunk 2 above)
Comment (jvican) [inline core/src/main/scala/scala/pickling/PicklingErrors.scala:75]: Cool, next PR will include these changes!

File: core/src/main/scala/scala/pickling/generator/sourcegen.scala (hunk 4 above)
Comment (jvican) [inline core/src/main/scala/scala/pickling/generator/sourcegen.scala:322]: Missed this one!

**Comment by jvican** [comment]
Thanks @jsuereth :smile: I'll address some stuff in my next PR. Currently, I'm trying to solve the scalajs problem and try these new changes with my custom picklers for spores! But I'll be doing this soon!
